File: elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  @methionine "AUG"
  @phenylalanine ["UUC", "UUU"]
  @leucine ["UUA", "UUG"]
  @serine ["UCU", "UCC", "UCA", "UCG"]
  @tyrosine ["UAU", "UAC"]
  @cysteine ["UGU", "UGC"]
  @tryptophan "UGG"
  @terminating ["UAA", "UAG", "UGA"]
  @stop "STOP"
  @codon_length ~r/.{3}/

  defguardp methionine?(codon) when codon == @methionine
  defguardp phenylalanine?(codon) when codon in @phenylalanine
  defguardp leucine?(codon) when codon in @leucine
  defguardp serine?(codon) when codon in @serine
  defguardp tyrosine?(codon) when codon in @tyrosine
  defguardp cysteine?(codon) when codon in @cysteine
  defguardp tryptophan?(codon) when codon == @tryptophan
  defguardp terminating?(codon) when codon in @terminating

  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    proteins =
      rna
      |> String.split(@codon_length, include_captures: true, trim: true)
      |> Enum.reduce([], &translate_codon/2)
      |> Enum.reverse()

    {:ok, proteins}
  catch
    {:stop, proteins} ->
      {:ok, proteins}

    :error ->
      {:error, "invalid RNA"}
  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon(codon) when methionine?(codon), do: {:ok, "Methionine"}
  def of_codon(codon) when phenylalanine?(codon), do: {:ok, "Phenylalanine"}
  def of_codon(codon) when leucine?(codon), do: {:ok, "Leucine"}
  def of_codon(codon) when serine?(codon), do: {:ok, "Serine"}
  def of_codon(codon) when tyrosine?(codon), do: {:ok, "Tyrosine"}
  def of_codon(codon) when cysteine?(codon), do: {:ok, "Cysteine"}
  def of_codon(codon) when tryptophan?(codon), do: {:ok, "Tryptophan"}
  def of_codon(codon) when terminating?(codon), do: {:ok, @stop}
  def of_codon(_codon), do: {:error, "invalid codon"}

  defp translate_codon(codon, acc) do
    codon
    |> of_codon()
    |> apply_protein(acc)
  end

  defp apply_protein({:ok, @stop}, acc), do: throw({:stop, Enum.reverse(acc)})
  defp apply_protein({:ok, protein}, acc), do: [protein | acc]
  defp apply_protein({:error, _message}, _acc), do: throw(:error)
end

File: typescript/protein-translation/protein-translation.ts
type Methionine = "AUG"
type Phenylalanine = "UUC" | "UUU"
type Leucine = "UUA" | "UUG"
type Serine = "UCU" | "UCC" | "UCA" | "UCG"
type Tyrosine = "UAU" | "UAC"
type Cysteine = "UGU" | "UGC"
type Tryptophan = "UGG"
type Terminating = "UAA" | "UAG" | "UGA"
type Codon =
  Methionine
  | Phenylalanine
  | Leucine
  | Serine
  | Tyrosine
  | Cysteine
  | Tryptophan
  | Terminating
type Stop = "STOP"
type Translation =
  "Methionine"
  | "Phenylalanine"
  | "Leucine"
  | "Serine"
  | "Tyrosine"
  | "Cysteine"
  | "Tryptophan"
  | Stop
type MaybeError<T> = T | never

export default class ProteinTranslation {
  private static readonly METHIONINE: Methionine = "AUG"
  private static readonly PHENYLALANINE: Readonly<Phenylalanine[]> =
    Object.freeze(["UUC", "UUU"])
  private static readonly LEUCINE: Readonly<Leucine[]> =
    Object.freeze(["UUA", "UUG"])
  private static readonly SERINE: Readonly<Serine[]> =
    Object.freeze(["UCU", "UCC", "UCA", "UCG"])
  private static readonly TYROSINE: Readonly<Tyrosine[]> =
    Object.freeze(["UAU", "UAC"])
  private static readonly CYSTEINE: Readonly<Cysteine[]> =
    Object.freeze(["UGU", "UGC"])
  private static readonly TRYPTOPHAN: Tryptophan = "UGG"
  private static readonly TERMINATING: Readonly<Terminating[]> =
    Object.freeze(["UAA", "UAG", "UGA"])
  private static readonly STOP: Stop = "STOP"
  private static readonly CODON: RegExp = /(.{3})/

  static proteins(rna: string): MaybeError<Translation[]> {
    try {
      return (
        rna
          .split(ProteinTranslation.CODON)
          .filter(Boolean)
          .reduce(ProteinTranslation.translateCodon, [])
      )
    } catch (retval) {
      if (retval instanceof Error) {
        throw retval
      }
      return retval
    }
  }

  private static translateCodon(
    acc: Translation[],
    codon: string
  ): Translation[] {
    const protein: Translation =
      ProteinTranslation.proteinForCodon(codon as Codon)

    if (protein === ProteinTranslation.STOP) {
      throw acc
    }

    acc.push(protein)
    return acc
  }

  private static proteinForCodon(codon: Codon): Translation {
    if (codon === ProteinTranslation.METHIONINE) {
      return "Methionine"
    } else if (
      ProteinTranslation.PHENYLALANINE.includes(codon as Phenylalanine)
    ) {
      return "Phenylalanine"
    } else if (ProteinTranslation.LEUCINE.includes(codon as Leucine)) {
      return "Leucine"
    } else if (ProteinTranslation.SERINE.includes(codon as Serine)) {
      return "Serine"
    } else if (ProteinTranslation.TYROSINE.includes(codon as Tyrosine)) {
      return "Tyrosine"
    } else if (ProteinTranslation.CYSTEINE.includes(codon as Cysteine)) {
      return "Cysteine"
    } else if (codon === ProteinTranslation.TRYPTOPHAN) {
      return "Tryptophan"
    } else {
      return ProteinTranslation.STOP
    }
  }
}
